smart_open/s3.py

```diff
@@ -587,28 +615,21 @@ def readable(self):
 
     def read(self, size=-1):
         """Read up to size bytes from the object and return them."""
-        if size == 0:
-            return b''
-        elif size < 0:
-            # call read() before setting _current_pos to make sure _content_length is set
-            out = self._read_from_buffer() + self._raw_reader.read()
-            self._current_pos = self._raw_reader._content_length
-            return out
 
-        #
-        # Return unused data first
-        #
-        if len(self._buffer) >= size:
-            return self._read_from_buffer(size)
+        # if an absolute size can be calculated we do calculate it to determine if it's available in the buffer already
+        if size < 0:
```

Comment:
It's better to avoid nesting conditionals where possible, for the sake of simplicity.

```python
if size < 0 and content_length is None:
   size = -1
elif size < 0:
   size = self._content_length - self._position
```